rfcs/20200211-tf-types.md

```diff
@@ -0,0 +1,295 @@
+
+```
+>>> class SupportsTensor(Protocol):
+...   def __tf_tensor__(self):
+...     pass
+>>> def f(x: SupportsTensor):
+...   pass
+>>> obj = MyClass()
+>>> f(obj)  # passes static type checks
+```
+
+Ultimately, `TensorLike` is to become a union of the protocol type along with any other types supported through legacy mechanisms:
+
+```
+TensorLike = Union[List, Tuple, tf.Tensor, SupportsTensor, ...]
+```
+
+The `Protocol` type is only standardized in Python 3.8. Backports exist through [typing_extensions](https://github.com/python/typing/tree/master/typing_extensions), although they still don’t support Python 3.5. Therefore, typing annotations will only be supported in 3.6+, and complete support is only available in 3.8+.
+
+Note that `Protocol` subtypes require [@runtime_checkable](https://www.python.org/dev/peps/pep-0544/#runtime-checkable-decorator-and-narrowing-types-by-isinstance) in order to be compatible with `isinstance`. However, that degrades the performance of `isinstance` in a way similar to `abc.ABCMeta`. For that reason, TensorFlow internal logic is encouraged to use the the more direct `hasattr` test for structural type checks of this kind.
+
+Although this RFC proposes the deprecation of `register_tensor_conversion_function`, it does not set a timeline for removing it. It remains an open question whether interim support for type annotations should be added to `register_tensor_conversion_function`.
+
+### Support for `tf.function`'s `input_signature`
+
+The type system listed here can be expanded to allow input signatures using type annotations, see for instance [this thread](https://github.com/tensorflow/tensorflow/issues/31579).
+
+Presently, the [input_signature](https://www.tensorflow.org/api_docs/python/tf/function) optional mechanism uses [tf.TensorSpec](https://www.tensorflow.org/api_docs/python/tf/TensorSpec) to describe the function arguments:
+
+```
+>>> @function(input_signature=[TensorSpec([3], dtype=int32)])
+... def f(x):
+...   tf.print(x)
+>>> f(constant([1, 2, 3]))
+[1 2 3]
+>>> f(constant([1, 2]))  # Shape mismatch
+ValueError: Python inputs incompatible with input_signature
+>>> f(constant([1.0, 2.0, 3.0]))  # DType mismatch
+ValueError: Python inputs incompatible with input_signature
+```
+
+It is expected however that some or all of this information will be repeated by the function's type annotations. Type annotations may be generic, for example by only specifying a dtype:
+
+```
+>>> @function(input_signature=[TensorSpec([3], dtype=int32)])
+... def f(x: Tensor[int32]):
+...   ...
+```
+
+In such cases, `tf.function` is expected to verify that the type annotation matches the `input_signature`.
+
+In the long term, this RFC recommends that type annotations fully replace the `input_signature`, so far as the Python type annotation system allows it:
+
+```
+>>> class BatchSize(Dimension):
+...   value = 32
+>>> @function
+... def f(x: Tensor[Int32, Shape1D[BatchSize]]):
+...   ...
```

Comment:
Why Shape1D rather than Shape? Could we also just use a tuple to specify the shape? It would lead to clearer signatures.

Reply:
I updated the RFC test to avoid prescribing a certain scheme for type annotations, and use whatever works best when we move to implementing it. It would be great to be able to use a tuple and we should do it if it's possible. In the few tests that I'm aware of [1], certain limitations in the type annotation system precluded the use of tuples, and forced using different type for different tensor ranks (e.g. Tensor1D, Tensor2D, etc.).

[1] https://github.com/tensorflow/tensorflow/issues/31579

Reply:
A plain tuple would be much clearer indeed. Perhaps an abstract `rank` as well in case the rank is unknown (e.g. `Shape(rank=None)` would be a dynamic-rank shape.

Reply:
The problem is `Tensor1D`, `Tensor2D` is that it hard-codes the rank, which isn't always a safe assumption. Could there be a type to specify "rank is dynamic but last dimension is N"? (which is what a `Dense` layer would need, for instance).

Reply:
The inability to mix different ranks is an unfortunate limitation of the type annotation system, as far as I can tell. But specifying an unknown rank or partial shape should be possible.

Reply:
> The inability to mix different ranks is an unfortunate limitation of the type annotation system, as far as I can tell. 

Could you expand on this? It would be desirable to have something as clear as [typing.Callable](https://docs.python.org/3/library/typing.html#callable). 

If `typing.Callable` can accept
```python
Callable[[str, int], float]
Callable[[None, int, int], Any]
Callable[..., int]
```

What would prevents us from having:

```python
# array of dim 3 or more, dim N°2 is unspecified.
def some_function(x: Tensor[Union[tf.float32, tf.float64], [32, 32, None, ...]]) -> tf.Tensor:
    pass

# array with unknown shape:
def sum(x: Tensor[tf.float32, None], y: Tensor[tf.float32, None]) -> tf.Tensor:
    return x + y

# array with any dtype
def sum(x: Tensor[Any, [32, 32]], y: Tensor[Any, [32, 32]]) -> tf.Tensor:
     return x + y
```

For the `...` we should follow numpy's conventions as much as possible of course.

Reply:
Interesting, that might work. Poking inside the source code, `Callable` and `Tuple` are special cases of the private `_VariadicGenericAlias`. I think that's what we'd need: https://github.com/python/cpython/blob/3.8/Lib/typing.py#L795

Anyway, the main requirement is contravariance, I think: and value of type `Tensor[Any, [32]]` should be compatible with an argument of type `Tensor[Any, [None]]`. At the same time, `Tensor[Any, [32, 32]]` should not be compatible `Tensor[Any, [None]]` but should be compatible with `Tensor[Any, None]`. How would the type declarations that allow us to to that look like?

Reply:
>  How would the type declarations that allow us to to that look like?

I'm not sure I understand your question. Are you talking about the API or the implementation details?

Reply:
I'm asking for a prototype which shows that e.g. `Tensor[tf.float32, [32, 32, None, ...]]` can be implemented.

Reply:
If we refer to the base implementation of tensorflow/tensorflow#31579 , currently it's not possible due to a limitation of the `input_signature` argument of tf.function which is too strict. For example, it's not possible to set a tf.TensorSpec which can accept multiple datatypes or specify only part of the shape (e.g. it's impossible to specify `[None, ..., 32]` as a valid shape). But there is no real reason why it shouldn't be possible to have a nice API like this. Tensorflow  would be free to redraw the graph if needed like it does when there is no `input_signature` to a tf.function.

Currently, setting an `input_signature` forces a python function to have only a single graph. But it's a big limitation. For example, it's impossible to do:

```python
# this function should make a new graph automatically for each new tensor dtype. 
# So 2 graphs max.
@tf.tfunction(input_signature=(tf.TensorSpec(shape=None, dtype=[tf.float32, tf.float64]),))
def square(x):
    return x**2
```

It's also impossible to specify a python object in a `input_signature` and arbitrary shapes too.

But it's needed badly. We have this issue in tensorflow addons where we can't use tf.function in our API because either it will draw too many graphs, either it will draw a single graph (which is not enough. e.g. multiple possible dtypes for tensors). 

The simplest use case is that you have a public API accepting in the same signature python object (which shouldn't be converted to tensors) and tensors with multiple data types, this happens frequently in tensorflow addons. We just can't use tf.function then.

The bottom line is that we need to change the tf.function API to have better control over the signature (python types, multiple dtype, any dimension for a tensor...), or else, we won't be able to implement and elegant API for type hints with tensorflow.

One possibility would be to go for your implementation proposed in this RFC, and when `input_signature` is improved and gets more flexible, we'll be allowed to use a clearer API and deprecate the old one.

The second possibility is to postpone the implementation of the type hints with tf.function until `input_signature` has more flexibility.

@qlzh727 , if I'm not mistaken, you're familiar with the implementation of tf.function. Do you think improving `input_signature` to allow more flexible signatures is possible?

Reply:
The RFC definitely doesn't finalize the design for this, so we have time - I'll clarify that in the doc.

I'm not too worried about the input_signature - we can add whatever functionality is missing to align it with type annotations, as long as it remains backward compatible. Backward compatibility isn't hard to achieve I think. The hard limitation is the amount of expressivity that the type annotations allow us, because they're not generic expressions and do have significant limitations. That's why I keep asking whether proposed interfaces are actually realizable in practice.

Reply:
I'm not familiar enough with the usage of `typing.Generic` to affirm anything concerning the faisability of this :( I think it's possible based on what's available in the standard library, but I'm not positive.

Reply:
With regards to syntax for shape typing, the broader Python ecosystem is very interested in this topic. Hopefully we can figure out a design/syntax that works across a range of array libraries (TensorFlow/NumPy/JAX/PyTorch) and type checkers (mypy/pytype/...).

Here are links to a few related discussions in other projects:
https://github.com/python/typing/issues/516
https://github.com/python/typing/issues/513
https://github.com/numpy/numpy/issues/7370

I would also suggest reaching out to the Python [typing-sig](https://mail.python.org/mailman3/lists/typing-sig.python.org/) mailing list, where this has also been a topic of repeated interest.

Reply:
Thank for the pointers, yes these are interesting discussions! I think only a few pieces are missing - happy to join the choir.

Reply:
I will add also https://github.com/google/jax/issues/943
https://github.com/google/jax/pull/2644
@shoyer check also this thread https://llvm.discourse.group/t/tcp-discussion-at-the-mlir-open-meeting-on-4-9/858/